### ISSUE TITLE
Weight Pitch Charts Based on Frequency

### ISSUE THREAD

**Issue by kruser** [kruser/pitchfx-site#31]
These pitch charts are reality...
![image](https://f.cloud.github.com/assets/2208700/2211672/e225c75e-99a8-11e3-9e80-744d7a558589.png)

...but the value in looking at them comes more from weighting based on frequency. That is, of course each player is going to show more green in the strike zone, because they see a heavy ratio of pitches there. The same is true for whiffs.

To get over this, we need to weight the opacity of bubbles according to the percentage of pitches that were seen in that zone.

Also, for whiffs, let's increase the radius of the bubble overall. As it stands we're not seeing a lot of overlap.

**Comment by kruser** [comment]
Highcharts is coming out with a highmaps library that includes heat maps. I'll probably replace these two charts with heat maps.

http://www.highcharts.com/maps/demo/heatmap

**Comment by kruser** [comment]
This heatmap library is a little more responsive to scatter data. 
http://www.patrick-wied.at/static/heatmapjs/examples.html

The highcharts implementation requires a value in each square.

**Comment by kruser** [comment]
Anybody care to offer an opinion on how you'd like to see hit zones? Choice in heat 
- http://www.highcharts.com/maps/demo/heatmap - with this one we'd be able to add tooltips with raw numbers
- http://www.patrick-wied.at/static/heatmapjs/examples.html - this one might look cooler

Either way, the charts will reflect the % of pitches in each zone. I'm thinking the zones should be represented by a 15x15 grid, with the inner 9x9 being the strike zone.

**Comment by albertlyu** [comment]
Correct me if I'm wrong, but heatmap.js will paint the heat in each zone relative to the rest of the heatmap, no? At least the example of http://www.patrick-wied.at/static/heatmapjs/example-heatmap-click.html, the max value that paints 'red' continues to increase if you click in one spot at a very high frequency. Perhaps the upper bound of the heat can be capped (% of pitches in each zone), but just my initial thought when exploring the demos.

**Comment by kruser** [comment]
@albertlyu, I'm not sure of the answer to that, but from a glance it looks like they support a max property on the configuration. The click demo doesn't have that property.

**Comment by kruser** [comment]
Also for this one, right now we just have the two zone charts....
- whiffs
- hit quality (wOBA)

I think we should add more like...
- swings
- whiffs per swings
- trajectory (grounders, flyballs, etc)

...and let the user scroll/toggle/click through to the one they're interested in.

I'm less interested in..
- BA
- Slugging
- HR

..as I think the wOBA chart kind of handles them all together.

**Comment by kruser** [comment]
Current plan for this..

![image](https://f.cloud.github.com/assets/2208700/2410315/0216c030-aabc-11e3-8ed3-dc5496901be4.png)

This is a mockup - the final product will be a little more polished.

**Comment by albertlyu** [comment]
Is this heatmap.js? This looks great. Some feedback: For the heatmaps, it would be great if we could show a legend for the colors (you may have already planned that, but just FYI). Also that legend color bar should hopefully have the same scale. Either way, I'd like to understand the code when you check this in, like how the heat classification is getting binned and whether there are any smoothing parameters, things like that. In my experience, heatmaps are definitely great, but we have to take care in how we display them so as not to mislead the user.

**Comment by kruser** [comment]
Yes, it is heatmap.js. I think you're right on the legend bar. I'll be sure to put that in. 

I think a v2 should include tooltips as well. Those will be a little more customized as heatmap.js doesn't support tooltips. Highcharts does, but I like the heatmap.js presentation better.

As for the scale, I agree. I'll have to put a scale that is the same for all players. Likely I'll base the scale on actual distributions, but hard code the bounds.

**Comment by kruser** [comment]
As much as I like the heatmap.js, the highcharts version has better support for tooltips, which will allow us to show data in each square, so I've switched gears to work with that.
http://www.highcharts.com/maps/demo/heatmap

The adjustments to the backend service are done, just making the UI changes now.

For now I'm doing a 10x10 grid, with the inner 6x6 grid being the common strike zone (not adjusted for batter height right now).

**Comment by albertlyu** [comment]
10x10 grid with a 6x6 grid inset? Are each of these squares going to be individual heat bins? I guess it's not as pretty as heatmap.js, but more functional.

**Comment by kruser** [comment]
Yes, each square would have its own value.

Example mockup without the colors
![image](https://f.cloud.github.com/assets/2208700/2488153/3607edea-b13f-11e3-93d9-1085a989c8e0.png)

**Comment by kruser** [comment]
I'm not sure if that's cutting it too thin or not. We'll see. I know brooksbaseball only does 5x5.

**Comment by albertlyu** [comment]
Ahh okay, I understand now. I like 6x6 inset because then it should be easier to aggregate and calculate values for the traditional 3x3 inset strikezone. In fact, letting the user play around with the bin sizes could be an option. Is that possible? Once a player page loads and grabs the values for each of the 100 bins, it should be relatively simple to quickly generate the values for a different combination of bins, as long as they're subset of the data, no?

I wonder what you think the use case for the end user would be to customize strikezones. For example, using your grid above:

![strikezone1](https://f.cloud.github.com/assets/1390356/2488215/98a533a8-b140-11e3-8c5b-cfc047e18ea4.png)

![strikezone2](https://f.cloud.github.com/assets/1390356/2488217/9bdb0750-b140-11e3-8671-a384b3c77587.png)

**Comment by albertlyu** [comment]
Hey @kruser, do I need to do any installations for the new highcharts to render the new heatmaps locally? I see the download link at highcharts.com, but I'm wondering if you think i should npm install highcharts or something along those lines?

**Comment by kruser** [comment]
@albertlyu , sorry about this, I've been building this feature in the master branch when I should have created a new branch. So for now you'll see empty slots as they haven't been implemented.

So far I have the Zones.js class that is responsible for moving pitches in the appropriate zones and then kicks out some stat for each zone.

You can see the test spec for Zones.js here. All tests are currently passing for whiffs/swings and swing % for each zone.
https://github.com/kruser/pitchfx-site/blob/master/app/test/spec/pojos/Zones.js

Hopefully this week I'll have the visuals in for those two charts. Then the rest of the charts should be pretty easy to fit in this framework.

**Comment by kruser** [comment]
This is coming along, although the colors are horrific right now. I'm not sure what to do when a 1/1 in a remote corner colors RED just like 100/100 in a square in the middle would. Anybody have any thoughts?

![image](https://f.cloud.github.com/assets/2208700/2528009/3d76a06a-b50b-11e3-8971-bbbab729bd46.png)

**Comment by kruser** [comment]
Another option here is to make the heatmap a 5x5 grid instead of 10x10 and display the %s and other data in each square - not in a tooltip. For example a square might show this...

75%
45/60

**Comment by albertlyu** [comment]
I'm not sure how user-friendly labeled data in each square would be. My suggestion is to use heatmap.js and add tool-tip features in the future. Are you sure there's no way to map a tooltip information on top of the strikezone sometime later? They don't have to be coordinated with heatmap.js, just have relative coordinates for each strikezone area and display specific tooltip / numbers on hover.

**Comment by kruser** [comment]
Tooltips are definitely feasible with heat map.JS, it's just not part of the API. 

I'm going to play a little more with heat map.JS and post an image to this thread. I think the Highcharts heatmap would look a lot better if we had 4x the squares. 10x10 looks more along the lines of the brooksbaseball zone charts.

**Comment by kruser** [comment]
Played with heatmap.js a little bit. It doesn't do great when the points are laid out in a grid. It would be a great way to plot out individual whiffs, hits, etc., like in a plain old scatter chart. However, the entire point of this issue is to make it so that we weight appropriately based on pitches seen.

![image](https://cloud.githubusercontent.com/assets/2208700/2557179/7ca09f5e-b6f3-11e3-9414-be83aa881b9c.png)

**Comment by kruser** [comment]
I moved back to the highmaps heatmap (highcharts). It displays grid data a little cleaner. Raw stats are available in the tooltip. Currently I only implemented swing rate and whiffs/swings, but it will be trivial to implement new metrics.

![image](https://cloud.githubusercontent.com/assets/2208700/2562537/8847126c-b850-11e3-9fcc-a8c2c0c161bf.png)

For those interested in adding more charts, start with looking into the tests for Zones, Zone and ZoneStat classes in this folder:
https://github.com/kruser/pitchfx-site/tree/master/app/app/scripts/pojos

New functions added to the Zones class can be retrieved and plotted fairly simply by adjusting the pitchStats.js controller.
